pybop/observers/unscented_kalman.py

```diff
@@ -0,0 +1,309 @@
+        """
+        Performs the unscented transform
+
+        Parameters
+        ----------
+        sigma_points : List[SigmaPoint]
+            The sigma points
+        sqrtR : np.ndarray
+            The square root of the covariance matrix
+
+        Returns
+        -------
+        Tuple[np.ndarray, np.ndarray]
+            The mean and covariance of the sigma points
+        """
+        x = np.sum(w_m * sigma_points, axis=1).reshape(-1, 1)
+        A = np.sqrt(w_c[1:]) * (sigma_points[:, 1:] - x)
+        (
+            _,
+            S,
+        ) = linalg.qr(A.T, mode="economic")
+        S = UkfFilter.cholupdate(S, sigma_points[:, 0:1] - x, w_c[0])
+        return x, S
+
+    @staticmethod
+    def cholupdate(R: np.ndarray, x: np.ndarray, w: float) -> np.ndarray:
+        """
+        Updates the cholesky decomposition of a matrix (see https://github.com/modusdatascience/choldate/blob/master/choldate/_choldate.pyx)
+
+        Note: will be in scipy soon so replace with this: https://github.com/scipy/scipy/pull/16499
+
+        TODO: need to replace with something more low-level
+
+        Parameters
+        ----------
+        R : np.ndarray
+            The cholesky decomposition of the matrix
+        x : np.ndarray
+            The vector to add to the matrix
+        w : float
+            The weight of the vector
+
+        Returns
+        -------
+        np.ndarray
+            The updated cholesky decomposition
+        """
+        x = x.flatten()
+        p = x.shape[0]
+        for k in range(p):
+            r = np.sqrt(R[k, k] ** 2 + w * x[k] ** 2)
+            # r = UkfFilter.hypot(R[k, k], x[k])
+            c = r / R[k, k]
+            s = x[k] / R[k, k]
+            R[k, k] = r
+            if k < p - 1:
+                R[k, k + 1 :] = (R[k, k + 1 :] + w * s * x[k + 1 :]) / c
+                x[k + 1 :] = c * x[k + 1 :] - s * R[k, k + 1 :]
+        return R
+
+    def step(self, y: np.ndarray) -> float:
+        """
+        Steps the filter forward one step using a measurement. Returns the log likelihood of the measurement.
+
+        Parameters
+        ----------
+        y : np.ndarray
+            The measurement vector
+
+        Returns
+        -------
+        float
+            The log likelihood of the measurement
+        """
+        sigma_points, w_m, w_c = self.gen_sigma_points(
+            self.x, self.S, self.alpha, self.beta
+        )
+        sigma_points = np.apply_along_axis(self.f, 0, sigma_points)
+
+        x_minus, S_minus = self.unscented_transform(sigma_points, w_m, w_c, self.sqrtRp)
+        sigma_points_y = np.apply_along_axis(self.h, 0, sigma_points)
+        y_minus, S_y = self.unscented_transform(sigma_points_y, w_m, w_c, self.sqrtRm)
+        P = np.sum(
+            w_c * np.multiply.outer(sigma_points - x_minus, sigma_points_y - y_minus),
+            axis=(1, 3),
+        )
+        gain = linalg.lstsq(linalg.lstsq(P.T, S_y.transpose())[0].T, S_y)[0]
```

Comment:
Did you try `linalg.solve` in-place of `linalg.lstsq`? Just wondering if there is stability issues with `linalg.solve` that requires us to use lstsq.

Reply:
its an over-determined set of equations, you are basically trying to fit a gaussian to a (possibly very) non-gaussian distribution, so `solve` won't work